cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillContext.java

```diff
@@ -211,101 +228,304 @@ public String getUID() {
+    private final AtomicInteger instances;
+    private final AtomicInteger instanceId;
+    private final CompletableFuture<AtomicInteger> instancesFuture;
+    private final CompletableFuture<AtomicInteger> instanceIdFuture;
+    private final ExecutorService executor;
+
+    InstanceWatcherThread(PodInfo podInfo, ApiClient apiClient, DynamicKubernetesApi api,
+                          V1OwnerReference ownerRef, CompletableFuture<AtomicInteger> instancesFuture,
+                          CompletableFuture<AtomicInteger> instanceIdFuture) {
+      super("twill-instance-watch", podInfo.getNamespace());
+      this.podInfo = podInfo;
+      this.apiClient = apiClient;
+      this.api = api;
+      this.ownerRef = ownerRef;
+      this.instances = new AtomicInteger();
+      this.instanceId = new AtomicInteger();
+      this.instancesFuture = instancesFuture;
+      this.instanceIdFuture = instanceIdFuture;
+      this.executor = Executors.newSingleThreadExecutor(Threads.createDaemonThreadFactory("twill-instance-assign"));
+    }
+
+    @Override
+    public void close() {
+      super.close();
+      executor.shutdownNow();
+    }
+
+    @Override
+    protected Watchable<DynamicKubernetesObject> createWatchable(Type resourceType, String namespace,
+                                                                 @Nullable String labelSelector) throws ApiException {
+      ListOptions listOptions = new ListOptions();
+      if (labelSelector != null) {
+        listOptions.setLabelSelector(labelSelector);
       }
+      listOptions.setFieldSelector("metadata.name=" + ownerRef.getName());
+      return api.watch(namespace, listOptions);
+    }
 
-      @Override
-      protected ApiClient getApiClient() {
-        return client;
+    @Override
+    public void resourceAdded(DynamicKubernetesObject resource) {
+      resourceModified(resource);
+    }
+
+    @Override
+    public void resourceModified(DynamicKubernetesObject resource) {
+      // For Job type, the number of instances is in the spec.parallelism field, otherwise it is in
+      // the status.replicas field
+      int count;
+      if ("Job".equals(ownerRef.getKind())) {
+        JsonObject spec = resource.getRaw().getAsJsonObject("spec");
+        if (spec == null) {
+          return;
+        }
+        count = spec.getAsJsonPrimitive("parallelism").getAsInt();
+      } else {
+        JsonObject status = resource.getRaw().getAsJsonObject("status");
+        if (status == null) {
+          return;
+        }
+        JsonPrimitive value = status.getAsJsonPrimitive("replicas");
+        if (value == null) {
+          return;
+        }
+        count = value.getAsInt();
       }
 
-      @Override
-      public void resourceAdded(Object resource) {
-        resourceModified(resource);
+      int oldInstances = instances.getAndSet(count);
+      if (oldInstances != count) {
+        LOG.debug("Number of instances change to {}", count);
```

Comment:
should we add `Number of instances changed from {} to {}`?

Reply:
Sure